api/swagger-batch.yaml

```diff
@@ -0,0 +1,310 @@
+openapi: 3.0.3
+info:
+  title: Turing Batch Experiment API
+  version: 0.0.1
+servers:
+  - url: localhost:8080/v1
+paths:
+  "/projects/{project_id}/ensemblers":
+    get:
+      tags: [ "Ensembler" ]
+      summary: Returns a list of ensemblers that belong to the project
+      parameters:
+        - in: path
+          name: project_id
+          schema:
+            $ref: "#/components/schemas/Id"
+          required: true
+        - in: query
+          name: page
+          schema:
+            type: integer
+            default: 1
+        - in: query
+          name: page_size
+          schema:
+            type: integer
+            default: 10
+      responses:
+        "200":
+          description: A JSON object, that represents paginated results response
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: "#/components/schemas/PaginatedResults"
+                  - type: object
+                    properties:
+                      results:
+                        type: array
+                        items:
+                          $ref: "#/components/schemas/GenericEnsembler"
```

Comment:
1. I don't think that this is required from the user-flow perspective (i.e. there would probably a view on UI for listing all ensemblers, so it would only render common properties of them).
2. It's not easy to implement this in gorm (or more accurately, generally in PQ), because you can't select from Parent AND child tables in a single query (you can't do a UNION of tables with a different column structure as well)

Reply:
I'm not sure if we should standardise the types of ensemblers returned here.

Right now the problem seems to be the GET and POST method seems to support different ensemblers. GET giving generic but POST only allowing pyfunc ensemblers. I think it would be easier in the future to support a field called `type` and every endpoint returning a GenericEnsembler type with a field called `metadata` that contains the specific metadata inside.

I might not be getting the pattern here but it seems difficult to extend this API to support other types of ensemblers in the future. I am guessing thats the reason why we are doing generic + subtype ensemblers here.

Reply:
I'm not too familar with the gorm library but i think it can be easily done with some if else statements to select from which tables depending on the type of ensembler.

Reply:
> I might not be getting the pattern here but it seems difficult to extend this API to support other types of ensemblers in the future.

@ashwinath I think the idea is to keep the `oneOf` updated with new types of ensemblers (but Roman can CMIIW). I echo your concern on the inconsistency between the endpoints.

@romanwozniak 
1. Would the UI need to show all types of ensemblers together? Or can we perhaps add a required `type` filter to the query?
2. I thought it could be multiple queries - one for each concrete type of ensemblers that is combined on the Golang side. But then the ordering, pagination, etc. are an issue. 🤔 I think if we really want to, it can still be done (query the generic table first and issue subsequent queries to the individual tables and and combine programmatically). But it is more complex.

Now thinking about it more, why are we not following the single table approach we've been using for the real-time Ensemblers (or the ExperimentEngine model)?
```
type: ...
standard_config: ...
docker_config: ...
```

Reply:
> Right now the problem seems to be the GET and POST method seems to support different ensemblers

@ashwinath I'm probably not getting the problem you are describing here. GET method (`projects/:project_id/ensemblers/:ensembler_id`) returns exactly the same data as was written into the database from POST `projects/:project_id/ensemblers`. GET `projects/:project_id/ensemblers` returns only generic data of each ensembler, but that is fine too, because the flow would be `List all ensemblers` -> `Click on a specific ensembler` -> `See the details of that ensembler` and List view would only need generic properties of each ensembler.

Reply:
> Now thinking about it more, why are we not following the single table approach we've been using for the real-time Ensemblers (or the ExperimentEngine model)?

@krithika369 I considered it, but preferred the table inheritance approach because it maps more naturally to the types composition in golang (i.e., I like the idea of having a `PyFuncEnsembler` struct with all the data available in it, instead of having a struct `Ensembler` and then look for `pyfunc_config` in it to check the data). It's probably possible to achieve the same with a single table approach and custom DB Write/Read methods, but that seems more complicated